### ISSUE TITLE
Massive slow down after 114455 nodes in heap, during extractMinimum()/delete() calls

### ISSUE THREAD

**Issue by Alex777UAE** [gwtw/js-fibonacci-heap#6]
It must be something with nodejs / v8 specific settings.

Problem could be easily repeated:

```
'use strict';

const FibonacciHeap = require('@tyriar/fibonacci-heap');
const fh = new FibonacciHeap((a, b) => {
    if (a.value.priority > b.value.priority) {
        return -1;
    }
    if (a.value.priority < b.value.priority) {
        return 1;
    }
    return 0;
});

console.log(process.memoryUsage());
console.log(new Date());

for (let i = 0; i < 114455; i++) {
    const name = 'A'+i;
    fh.insert('A'+i, {priority: i, name, test: [54.1323, 76.9999223]});
}
console.log(new Date());
const min = fh.findMinimum();
console.log(min.value.name);
fh.delete(min);
console.log(new Date());
console.log(process.memoryUsage());
console.log(new Date());
```

You'll see how slow it became if you set 114455 to be 114456.

Could you please advice what could be done to leverage this problem?

**Comment by Tyriar** [comment]
This actually makes sense. The [`extractMinimum`](http://www.growingwiththeweb.com/2014/06/fibonacci-heap.html#extract-minimum) operation (and `delete` since it calls `extractMinimum`) has a time complexity of _O(log n) [amortized](http://stackoverflow.com/a/15079679/1156119)_, it's actual worst case is `O(n)`. One of the ways the FIbonacci heap achieves it's good (amortized) run times is to defer "consolidating" nodes inserted via the [_insert_](http://www.growingwiththeweb.com/2014/06/fibonacci-heap.html#insert) operation, "consolidate" is only called on `extractMinimum`.

Here's basically what the code is doing:
1. Add 114455 nodes to the tree, adding them as individual trees so that the tree looks something like this:
   
   ```
   x-x-x-x-x-x-x-x-x-x-x-x-x-....-x
   ```
2. Call delete node `A`
   1. Call decreaseKey on `A`, decreasing the key to negative infinity and pushing it to the top of the tree (it is already at the top of the single node tree)
   2. Call extractMinimum on `A` which "consolidates" the tree, recursively merging nodes into trees such that no tree of the same rank exists

2.ii. is the part that comes out as _O(n)_ since every single node in the Fibonacci Heap is a single node tree that must be merged. After 2.ii. the Fibonacci heap will look something like this, where the nodes at the top of the tree are the minimum nodes for that tree:

```
x--x----x--------x---(and so on)
   |   /|     /-/|
   x  x x    x x x
      x     /| |
           x x x  
           | 
           x  
```

You can read more about the data structure [here](http://www.growingwiththeweb.com/2014/06/fibonacci-heap.html).

**Comment by Alex777UAE** [comment]
Thanks for the reply, but I think it's not a case, correct me if I'm wrong.
Just try to run an example code with 114455 nodes first and 114456 then.
You will see how huge is the the difference in time taken between first and the second case.
I think it has nothing to do with the algorithm itself and data structure.

I mean 114455 is not just a random number, it's a boundary after wich we have some specific issues with nodejs (my guess).

**Comment by Tyriar** [comment]
I actually see around the same time:

```
❯ node index.js
{ rss: 26120192, heapTotal: 10522624, heapUsed: 4231320 }
2016-10-14T23:29:06.177Z
2016-10-14T23:29:06.283Z
A114454
2016-10-14T23:29:30.638Z
{ rss: 96411648, heapTotal: 77705216, heapUsed: 41089016 }
2016-10-14T23:29:30.638
```

```
❯ node index.js
{ rss: 26095616, heapTotal: 10522624, heapUsed: 4231320 }
2016-10-14T23:51:05.940Z
2016-10-14T23:51:06.033Z
A114455
2016-10-14T23:51:29.332Z
{ rss: 96387072, heapTotal: 77705216, heapUsed: 41078136 }
2016-10-14T23:51:29.333Z
```

If you do observe a significant slow down I'd also guess it's something to do with the environment you're running it in.

**Comment by Alex777UAE** [comment]
Could you please advice me on nodejs version you are using?

I tried it on mac os sierra 10.2 & ubuntu 14.04
with node 5.4.11 and 6.8.0

**Comment by Tyriar** [comment]
Ubuntu 16.04, node v6.8.0

**Comment by Tyriar** [comment]
For good measure I approximately doubled the number and found that the run time increased by quite a bit (~25s vs ~90s):

```
❯ node index.js
{ rss: 26177536, heapTotal: 10522624, heapUsed: 4231320 }
2016-10-14T23:55:53.635Z
2016-10-14T23:55:53.812Z
A224456
2016-10-14T23:57:24.627Z
{ rss: 131342336, heapTotal: 112152576, heapUsed: 75368536 }
2016-10-14T23:57:24.627Z
```

I think this is expected though as the order of the trees will progressively get larger and larger as the node count increases. It should still be O(n) though.

**Comment by Alex777UAE** [comment]
Now I'm getting this problem with another "not random number" - 71109 vs 71110
Let me demonstrate it:

{ rss: 25436160, heapTotal: 10522624, heapUsed: 4232216 }
2016-10-15T00:04:30.321Z
2016-10-15T00:04:30.396Z
A71108
2016-10-15T00:04:30.420Z
{ rss: 60428288, heapTotal: 56659968, heapUsed: 28132656 }
2016-10-15T00:04:30.421Z

it took 24ms for 71109

{ rss: 25206784, heapTotal: 10522624, heapUsed: 4232216 }
2016-10-15T00:05:39.790Z
2016-10-15T00:05:39.867Z
A71109
2016-10-15T00:05:47.918Z
{ rss: 61206528, heapTotal: 57532416, heapUsed: 28132912 }
2016-10-15T00:05:47.918Z

and for 71110 it takes more than 8 seconds

**Comment by Alex777UAE** [comment]
It was on nodejs 6.8.0 already, btw.
I assure, there is no other CPU-heavy processes in system interfering with the result.
It might be something connected with nodejs RAM management. May be new space vs old space.
I'll investigate it more.

**Comment by Alex777UAE** [comment]
The problem starts with

v8::internal::Heap::MoveElements(v8::internal::FixedArray*, int, int, int)

Here is processed profiling of "71109" nodes heap:

>  [Shared libraries]:
>    ticks  total  nonlib   name
>       2    0.5%          /usr/lib/system/libsystem_pthread.dylib
> 
>  [JavaScript]:
>    ticks  total  nonlib   name
>      11    2.6%    2.6%  LazyCompile: *<anonymous> /Users/alex/Projects/untitled1/fibonacci.js:1:11
>       8    1.9%    1.9%  LazyCompile: *consolidate /Users/alex/Projects/untitled1/node_modules/@tyriar/fibonacci-heap/index.js:264:21
>       5    1.2%    1.2%  Stub: StringAddStub_ConvertRight_NotTenured
>       2    0.5%    0.5%  Stub: RecordWriteStub {3}
>       2    0.5%    0.5%  Stub: CEntryStub {1}
>       2    0.5%    0.5%  LazyCompile: *NodeListIterator /Users/alex/Projects/untitled1/node_modules/@tyriar/fibonacci-heap/index.js:184:33
>       2    0.5%    0.5%  Function: ~<anonymous> /Users/alex/Projects/untitled1/fibonacci.js:1:11
>       1    0.2%    0.2%  Stub: ToBooleanStub
>       1    0.2%    0.2%  Stub: RecordWriteStub {6}
>       1    0.2%    0.2%  Stub: RecordWriteStub {5}
>       1    0.2%    0.2%  Stub: RecordWriteStub {4}
>       1    0.2%    0.2%  Stub: RecordWriteStub {2}
>       1    0.2%    0.2%  Stub: RecordWriteStub {1}
>       1    0.2%    0.2%  Stub: RecordWriteStub
>       1    0.2%    0.2%  Stub: FastNewClosureStub
>       1    0.2%    0.2%  Stub: CreateWeakCellStub
>       1    0.2%    0.2%  Stub: CompareICStub
>       1    0.2%    0.2%  Stub: CEntryStub
>       1    0.2%    0.2%  Stub: BinaryOpWithAllocationSiteStub
>       1    0.2%    0.2%  LazyCompile: ~slice buffer.js:805:40
>       1    0.2%    0.2%  LazyCompile: *FibonacciHeap.insert /Users/alex/Projects/untitled1/node_modules/@tyriar/fibonacci-heap/index.js:124:43
>       1    0.2%    0.2%  LazyCompile: *FibonacciHeap /Users/alex/Projects/untitled1/fibonacci.js:6:30
>       1    0.2%    0.2%  KeyedLoadIC: A keyed load IC from the snapshot
>       1    0.2%    0.2%  Builtin: ArrayShift
> 
>  [C++]:
>    ticks  total  nonlib   name
>      58   13.9%   13.9%  node::ContextifyScript::New(v8::FunctionCallbackInfov8::Value const&)
>      37    8.9%    8.9%  ___mac_set_fd
>      12    2.9%    2.9%  v8::internal::IncrementalMarking::Step(long, v8::internal::IncrementalMarking::CompletionAction, v8::internal::IncrementalMarking::ForceMarkingAction, v8::internal::IncrementalMarking::ForceCompletionAction)
>      11    2.6%    2.6%  node::Binding(v8::FunctionCallbackInfov8::Value const&)
>       9    2.2%    2.2%  v8::internal::Heap::IteratePromotedObjectPointers(v8::internal::HeapObject_, unsigned char_, unsigned char_, bool, void (_)(v8::internal::HeapObject*_, v8::internal::HeapObject_))

And so on. Everything looks fine.

Now take a look at "71110" nodes profile:

>  [Shared libraries]:
>    ticks  total  nonlib   name
>       1    0.0%          /usr/lib/system/libsystem_pthread.dylib
>       1    0.0%          /usr/lib/system/libsystem_c.dylib
> 
>  [JavaScript]:
>    ticks  total  nonlib   name
>      14    0.2%    0.2%  LazyCompile: *consolidate /Users/alex/Projects/untitled1/node_modules/@tyriar/fibonacci-heap/index.js:264:21
>       8    0.1%    0.1%  LazyCompile: *<anonymous> /Users/alex/Projects/untitled1/fibonacci.js:1:11
>       3    0.0%    0.0%  Stub: StringAddStub_ConvertRight_NotTenured
>       3    0.0%    0.0%  KeyedLoadIC: A keyed load IC from the snapshot
>       2    0.0%    0.0%  Stub: FastArrayPushStub
>       1    0.0%    0.0%  Stub: ToBooleanStub
>       1    0.0%    0.0%  Stub: RecordWriteStub
>       1    0.0%    0.0%  Stub: LoadFieldStub
>       1    0.0%    0.0%  Stub: FastNewObjectStub
>       1    0.0%    0.0%  Stub: CompareICStub
>       1    0.0%    0.0%  Stub: CEntryStub {1}
>       1    0.0%    0.0%  Stub: CEntryStub
>       1    0.0%    0.0%  Stub: BinaryOpWithAllocationSiteStub
>       1    0.0%    0.0%  Stub: AllocateInNewSpaceStub
>       1    0.0%    0.0%  LazyCompile: ~consolidate /Users/alex/Projects/untitled1/node_modules/@tyriar/fibonacci-heap/index.js:264:21
>       1    0.0%    0.0%  LazyCompile: ~NativeModule.exists bootstrap_node.js:447:33
>       1    0.0%    0.0%  LazyCompile: readPackage module.js:79:21
>       1    0.0%    0.0%  LazyCompile: *NodeListIterator /Users/alex/Projects/untitled1/node_modules/@tyriar/fibonacci-heap/index.js:184:33
>       1    0.0%    0.0%  LazyCompile: *FibonacciHeap.insert /Users/alex/Projects/untitled1/node_modules/@tyriar/fibonacci-heap/index.js:124:43
>       1    0.0%    0.0%  KeyedLoadIC: A keyed load IC from the snapshot {1}
>       1    0.0%    0.0%  Handler: An IC handler from the snapshot
>       1    0.0%    0.0%  Function: ~<anonymous> /Users/alex/Projects/untitled1/fibonacci.js:1:11
>       1    0.0%    0.0%  Builtin: CallFunction_ReceiverIsNotNullOrUndefined
>       1    0.0%    0.0%  Builtin: CallFunction_ReceiverIsAny
>       1    0.0%    0.0%  Builtin: ArrayShift
> 
>  [C++]:
>    ticks  total  nonlib   name
>    6454   83.4%   83.4%  v8::internal::Heap::MoveElements(v8::internal::FixedArray_, int, int, int)
>     894   11.6%   11.6%  __flush_reset
>      53    0.7%    0.7%  node::ContextifyScript::New(v8::FunctionCallbackInfov8::Value const&)
>      37    0.5%    0.5%  ___mac_set_fd
>       6    0.1%    0.1%  void v8::internal::BodyDescriptorBase::IterateBodyImplv8::internal::StaticScavengeVisitor(v8::internal::Heap_, v8::internal::HeapObject_, int, int)
>       6    0.1%    0.1%  void v8::internal::BodyDescriptorBase::IterateBodyImplv8::internal::IncrementalMarkingMarkingVisitor(v8::internal::Heap_, v8::internal::HeapObject*, int, int)

Do you have any idea what roots this change has?

**Comment by Alex777UAE** [comment]
Here is related topic, I think:

https://bugs.chromium.org/p/v8/issues/detail?id=2131

> I strongly suspect that this is due to the backing store of the larger array being allocated in large-object space, where we currently cannot do efficient trimming of arrays from the left. So we end up moving the whole content to the left and adding the slots to the store-buffer because the backing store is not in new-space. This is especially brain-dead in this case, where we actually do not move anything at all, because 'start' is set to zero. I have some ideas how to improve Array.prototype.splice for these cases.

**Comment by Tyriar** [comment]
Fixed with https://github.com/gwtw/js-fibonacci-heap/pull/7 and published 1.0.8, thanks!
